wsproto/connection.py

```diff
@@ -275,6 +275,11 @@ def events(self):
                 if frame.opcode is Opcode.PING:
                     assert frame.frame_finished and frame.message_finished
                     self._outgoing += self._proto.pong(frame.payload)
+                    yield PingReceived(frame.payload)
+
+                elif frame.opcode is Opcode.PONG:
+                    assert frame.frame_finished and frame.message_finished
+                    yield PongReceived(frame.payload)
```

Comment:
From RFC6455:
> It SHOULD respond with Pong frame as soon as is practical.

But it is also allowed to send PONGs, without receiving a PING first.

> A Pong frame sent in response to a Ping frame must have identical  "Application data" as found in the message body of the Ping frame being replied to.

Yes, I think we should store the ping payload, and verify it against the first pong we receive, and the reset it to None (or similar?).

If we want to overcomplicate this, we should introduce a stack that holds all ping-payloads, and we pop them off one-by-one for every pong received...

For now, I would just clearly document that every `ping(payload)` overwrites the "global" expected pong response payload. So, don't send a PING before you got the PONG of the last one back.